compose/service.py

```diff
@@ -685,15 +685,30 @@ def get_volumes_from_names(self):
     # TODO: this would benefit from github.com/docker/docker/pull/14699
     # to remove the need to inspect every container
     def _next_container_number(self, one_off=False):
-        containers = filter(None, [
-            Container.from_ps(self.client, container)
-            for container in self.client.containers(
-                all=True,
-                filters={'label': self.labels(one_off=one_off)})
-        ])
+        containers = self._fetch_and_inspect(
+            all=True,
+            filters={'label': self.labels(one_off=one_off)}
+        )
         numbers = [c.number for c in containers]
         return 1 if not numbers else max(numbers) + 1
 
+    def _fetch_and_inspect(self, **fetch_options):
+        # Account for containers that might have been removed since we fetched
+        # the list.
+        def soft_inspect(container):
+            if container is None:
+                return None
+            try:
```

Comment:
SInce we're calling inspect anyway, the following can be shortened thus:
```python
try:
  return Container.from_id(self.client, container['Id'])
except NotFound:
  return None
```

and the `soft_inspect` call modified accordingly to use the raw container dict.